### PR TITLE
Meta: enable more lint rules

### PR DESCRIPTION
**Pull request by yakov116**
@fregante I went through them all only `@typescript-eslint/require-await': 'error'` am I not sure if they are correct.

### CODE REVIEW

File: source/features/add-tag-to-commits.tsx [hunk 1]
```diff
@@ -134,9 +134,6 @@ async function init(): Promise<void | false> {
 			// No tags for this commit found in the cache, check in github
 			cached = mergeTags(cached, await getTags(lastCommitOnPage)); // eslint-disable-line no-await-in-loop
 			targetTags = cached[targetCommit];
-		}
-
-		if (!targetTags) {
```
Comment (yakov116) [inline source/features/add-tag-to-commits.tsx:139]: Please Double check that my change is correct and for some reason it was designed like this.

Comment (fregante) [reply source/features/add-tag-to-commits.tsx:139]: No, this should be undone, as you can see it tries to set targetTags in 2 ways, so it has to check twice.

File: source/features/remove-label-faster.tsx [hunk 2]
```diff
@@ -36,7 +36,7 @@ async function removeLabelButtonClickHandler(event: delegate.Event<MouseEvent, H
 }
 
 // TODO: set variable via JSX and inline function in `init` after https://github.com/vadimdemedes/dom-chef/issues/66
-function makeRemoveLabelButton(labelName: string, backgroundColor: string) {
+function makeRemoveLabelButton(labelName: string, backgroundColor: string): HTMLButtonElement {
```
Comment (fregante) [inline source/features/remove-label-faster.tsx:39]: Unless we need a specific tag (e.g. to access href), just use `JSX.Element` as the type, so you don’t have to use the assertions below

File: source/features/remove-projects-tab.tsx [hunk 3]
```diff
@@ -7,7 +7,7 @@ import * as pageDetect from 'github-url-detection';
 import features from '.';
 import {observeOneMutation} from '../helpers/simplified-element-observer';
 
-function getProjectsTab() {
+async function getProjectsTab(): Promise<HTMLElement | undefined> {
```
Comment (fregante) [inline source/features/remove-projects-tab.tsx:10]: I’m not particularly fond of the rule that enforces this (“promise functions must be async). If it’s the only case where this happens, fine, but if it requires a lot of similar unnecessary changes let’s remove it

Comment (fregante) [reply source/features/remove-projects-tab.tsx:10]: Ok, good as is.

**Comment by fregante** [comment]
Everything else looks good so far. Let me know when you’re done making changes, I might have other rule suggestions

**Comment by yakov116** [comment]
I am good except for my first comment

**Comment by fregante** [comment]
Ok, can you sort them?

1. Group by state (2 groups: either disabled from the shared config, or new rule)
2. Alphabetically in each group

```js
rules: {
	/* Disable rules */
	a: "off",
	b: "off",

	/* Add rules */
	c: "error",
	d: "error",
}
```

File: .xo-config.js [hunk 4]
```diff
@@ -44,7 +44,28 @@ module.exports = {
 		// Temporarily disabled
 		// TODO: Need to decide what to do with this rule. Either keep it disabled here or remove it from XO.
 		'@typescript-eslint/no-invalid-void-type': 'off', // Reason: https://github.com/sindresorhus/refined-github/pull/3168#discussion_r434479434
-		'unicorn/no-fn-reference-in-iterator': 'off'
+		'unicorn/no-fn-reference-in-iterator': 'off',
+		'@typescript-eslint/no-extra-non-null-assertion': 'error',
+		'@typescript-eslint/consistent-type-definitions':'error',
+		'@typescript-eslint/require-await': 'error',
```
Comment (sindresorhus) [inline .xo-config.js:50]: FYI: https://github.com/xojs/eslint-config-xo-typescript/blob/7842d05deaadff1ced9a46f3e56d786e7b2922d2/index.js#L443

File: .xo-config.js [hunk 5]
```diff
@@ -44,7 +44,28 @@ module.exports = {
 		// Temporarily disabled
 		// TODO: Need to decide what to do with this rule. Either keep it disabled here or remove it from XO.
 		'@typescript-eslint/no-invalid-void-type': 'off', // Reason: https://github.com/sindresorhus/refined-github/pull/3168#discussion_r434479434
-		'unicorn/no-fn-reference-in-iterator': 'off'
+		'unicorn/no-fn-reference-in-iterator': 'off',
+		'@typescript-eslint/no-extra-non-null-assertion': 'error',
+		'@typescript-eslint/consistent-type-definitions':'error',
```
Comment (sindresorhus) [inline .xo-config.js:49]: While it might default to `interface`, this is not documented and also not clear.

Also, any reasoning to defaulting to `interface`?

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/consistent-type-definitions.md

The only benefit of `interface` is declaration merging, but `type` has other benefits too: https://stackoverflow.com/a/52682220

File: .xo-config.js [hunk 6]
```diff
@@ -44,7 +44,28 @@ module.exports = {
 		// Temporarily disabled
 		// TODO: Need to decide what to do with this rule. Either keep it disabled here or remove it from XO.
 		'@typescript-eslint/no-invalid-void-type': 'off', // Reason: https://github.com/sindresorhus/refined-github/pull/3168#discussion_r434479434
```
Comment (sindresorhus) [inline .xo-config.js:46]: This was disabled in the shared config.

File: .xo-config.js [hunk 7]
```diff
@@ -44,7 +44,28 @@ module.exports = {
...
+		'@typescript-eslint/explicit-function-return-type': [
+			'error',
+			{
+				allowExpressions: true,
+				allowTypedFunctionExpressions: true,
+				allowHigherOrderFunctions: true,
+				allowConciseArrowFunctionExpressionsStartingWithVoid: false
+			}
+		],
+		'@typescript-eslint/explicit-module-boundary-types': [
```
Comment (sindresorhus) [inline .xo-config.js:60]: Just make sure you're ok with this: https://github.com/xojs/eslint-config-xo-typescript/blob/7842d05deaadff1ced9a46f3e56d786e7b2922d2/index.js#L136-L138

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: .xo-config.js (hunk 7 above)
Comment (fregante) [reply .xo-config.js:60]: According to [its docs](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-module-boundary-types.md#when-not-to-use-it), this rule is only a "less strict" version of `explicit-function-return-type`, which we already use. Right?

If so, it should be dropped from here and from XO’s comments

File: .xo-config.js (hunk 5 above)
Comment (fregante) [reply .xo-config.js:49]: > but `type` has other benefits too

The only benefit I see is that you can use primitives, like `type S = string`, but this rule doesn't seem to prevent us from using that.

From the replacements I see in this PR, it only affects `type`s that can as well be `interface`s.

In my eyes, the difference between the two:

- `type x = {};` looks like a declaration
- `interface x {}` looks like a definition

Minimal difference, really, it's just about consistency

File: .xo-config.js (hunk 4 above)
Comment (fregante) [reply .xo-config.js:50]: I remember this. It looks like the lint is failing because of this one.

If it works painlessly on the current code, we can enable it.

File: .xo-config.js (hunk 5 above)
Comment (sindresorhus) [reply .xo-config.js:49]: > The only benefit I see is that you can use primitives, like type S = string, but this rule doesn't seem to prevent us from using that.

Sure, but if you ever need that, you're back to inconsistency. Which is why I slightly prefer `type`, but it doesn't matter much. 🤷‍♂️

Kinda annoying that TS has such similar duplicate things.

Comment (fregante) [reply .xo-config.js:49]: > any reasoning to defaulting to `interface`

I like definitions over declarations. The former looks cleaner to me:

```ts
interface X {}
type X = {};
```

Kinda like I prefer `class X {}` over `const X = class {}`

File: .xo-config.js (hunk 4 above)
Comment (yakov116) [reply .xo-config.js:50]: This was the only one I was having issues with (hence my first comment)

**Review by fregante — APPROVED** [review]

**Comment by fregante** [comment]
If you want to update the dependencies, this would be useful: https://github.com/fregante/webext-storage-cache/pull/20
